smack-extensions/src/main/java/org/jivesoftware/smackx/usertune/element/UserTuneElement.java

```diff
@@ -0,0 +1,246 @@
+import org.jivesoftware.smack.packet.XmlEnvironment;
+import org.jivesoftware.smack.util.NumberUtil;
+import org.jivesoftware.smack.util.XmlStringBuilder;
+
+/**
+ * {@link ExtensionElement} that contains the UserTune.
+ * <br>
+ * Instance of UserTuneElement can be created using {@link Builder#build()} method.
+ * To communicate a meaningful UserTune ,{@link #title} and {@link #source} are set as compulsory parameters.
+ */
+public final class UserTuneElement implements ExtensionElement {
+
+    public static final String NAMESPACE = "http://jabber.org/protocol/tune";
+    public static final String ELEMENT = "tune";
+
+    private final String artist;
+    private final Integer length;
+    private final Integer rating;
+    private final String source;
+    private final String title;
+    private final Integer track;
+    private final URI uri;
+
+    public static final UserTuneElement EMPTY_USER_TUNE = null;
+
+    private UserTuneElement() {
+        this(null);
+    }
+
+    private UserTuneElement(Builder builder) {
+        this.artist = builder.artist;
+        this.length = builder.length;
+        this.rating = builder.rating;
+        this.source = builder.source;
+        this.title = builder.title;
+        this.track = builder.track;
+        this.uri = builder.uri;
+    }
+
+    @Override
+    public String getNamespace() {
+        return NAMESPACE;
+    }
+
+    @Override
+    public String getElementName() {
+        return ELEMENT;
+    }
+
+    @Override
+    public XmlStringBuilder toXML(XmlEnvironment xmlEnvironment) {
+        XmlStringBuilder xml = new XmlStringBuilder(this);
+        if (isEmptyUserTune()) {
+            return xml.closeEmptyElement();
+        }
+        xml.rightAngleBracket();
+        xml.optAppend(getUserTuneXML());
+        return xml.closeElement(getElementName());
+    }
+
+    private CharSequence getUserTuneXML() {
+
+        XmlStringBuilder userTuneXML = new XmlStringBuilder();
+        userTuneXML.optElement("artist", artist);
+        userTuneXML.optElement("length", length);
+        userTuneXML.optElement("rating", rating);
+        userTuneXML.element("source", source);
+        userTuneXML.element("title", title);
+        userTuneXML.optElement("track", track);
+        userTuneXML.optElement("uri", uri);
+        return userTuneXML.toString();
+    }
+
+    private boolean isEmptyUserTune() {
+        if (this.artist == null && this.length == null && this.rating == null && this.source == null && this.title == null
+                && this.track == null && this.uri == null) {
+            return true;
+        }
+        return false;
+    }
+
+    public static boolean hasUserTuneElement(Message message) {
+        return message.hasExtension(ELEMENT, NAMESPACE);
+    }
+
+    public static UserTuneElement from(Message message) {
+        return message.getExtension(ELEMENT, NAMESPACE);
+    }
+
+    @Override
+    public int hashCode() {
+        return super.hashCode();
```

Comment:
That is wrong because it breaks the API contract of `hashCode()` and `equals()`.